PUBDEV-5654 H2OXGBoost result differ from native XGBoost when dmatrix…

Pscheidl opened the pull request:
…_type=sparse

https://0xdata.atlassian.net/browse/PUBDEV-5654

michalkurka commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java:
```diff
@@ -552,16 +552,16 @@ private static int initalizeFromChunkIds(Frame f, int[] chunks, Vec.Reader[] vec
 
                 for (int j = 0; j < di._cats; ++j) {
-                    if (!vecs[j].isNA(i)) {
+                    if (vecs[j].at(i) != 0) {
```
This seems incorrect. You are skipping values that have categorical level = 0. If categorical has level 2, you will correctly output (0, 0, 1), but if it has level 0 you will not output anything.

Pscheidl replied:
Yes, this is incorrect.

michalkurka commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java:
```diff
@@ -599,15 +599,15 @@ private static int initializeFromChunks(Chunk[] chunks, int weight, DataInfo di,
 
             for (int j = 0; j < di._cats; ++j) {
-                if (!chunks[j].isNA(i)) {
+                if (chunks[j].atd(i) != 0) {
```
Same issue as above - categorical and numerical values need to be treated differently.

wendycwong replied:
Remember to add the case when NA appears in chunks[j].atd(i).

Pscheidl replied:
I do it with at8 and an if-else statement.

michalkurka commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java:
```diff
@@ -599,15 +599,15 @@ private static int initializeFromChunks(Chunk[] chunks, int weight, DataInfo di,
+                if (chunks[j].atd(i) != 0) {
                     data[currentRow][currentCol] = 1; //one-hot encoding
-                    colIndex[currentRow][currentCol++] = di.getCategoricalId(j, chunks[j].at8(i));
+                    colIndex[currentRow][currentCol++] = di.getCategoricalId(j, chunks[j].atd(i));
```
why is this needed?

Pscheidl replied:
Unintentional refactor.

Pscheidl commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java:
```diff
@@ -552,16 +552,16 @@ private static int initalizeFromChunkIds(Frame f, int[] chunks, Vec.Reader[] vec
+                    if (vecs[j].at(i) != 0) {
                         data[currentRow][currentCol] = 1; //one-hot encoding
-                        colIndex[currentRow][currentCol++] = di.getCategoricalId(j, vecs[j].at8(i));
+                        colIndex[currentRow][currentCol++] = di.getCategoricalId(j, vecs[j].at(i));
```
Yes, will be. This is still WIP !

Pscheidl commented:
This is how array flattening in case of our CSR matrix works:

```c
/*
 * Class:     ml_dmlc_xgboost4j_java_XGBoostJNI
 * Method:    XGDMatrixCreateFrom2DCSREx
 * Signature: ([J[J[F)J
 */
JNIEXPORT jint JNICALL Java_ml_dmlc_xgboost4j_java_XGBoostJNI_XGDMatrixCreateFrom2DCSREx
  (JNIEnv *jenv, jclass jcls, jobjectArray jindptr, jobjectArray jindices, jobjectArray jdata, jint jcol, jint jrows, jlong total, jlongArray jout) {
  DMatrixHandle result;

  jlong* indptr = flatten_long(jenv, jindptr, jrows, 0);
  jint* indices = flatten_int(jenv, jindices, total, 0);
  jfloat* data = flatten_float(jenv, jdata, total, 0);

  bst_ulong nindptr = (bst_ulong)jrows;
  bst_ulong nelem = (bst_ulong)total;

  jint ret = (jint) XGDMatrixCreateFromCSREx((size_t const *)indptr, (unsigned int const *)indices, (float const *)data, nindptr, nelem, jcol, &result);

  setHandle(jenv, jout, result);
  delete[] indptr;
  delete[] indices;
  delete[] data;

  return ret;
}
```

Pscheidl commented:
Please note currentRow in ```initializeFromChunks``` is never incremented. Will never jump into a second row of data.

wendycwong commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java:
```diff
@@ -540,95 +533,90 @@ private static int initalizeFromChunkIds(Frame f, int[] chunks, Vec.Reader[] vec
                     data[currentRow][currentCol] = 0;
                     colIndex[currentRow][currentCol++] = 0;
-                    nz++;
+                    nonZeroCount++;
```
Not sure I get this part.  If the whole chunk is 0, then, it should be left alone and we save memory by not recording it.  There must be some other way to keep track of number of rows between the predictors and weight vectors.  I think python may not have done this....

Pscheidl replied:
This is not about whole chunk, but about whole row. If a whole row is zero-only, then there must be a record somewhere that there is an empty row (for proper reconstruction of the full matrix, if nothing else). Scipy duplicate row index, @mdymczyk 's API uses one more zero and also puts a row index into row headers. This API was there from the beginning and I don't want to break it.

It seems to be interpreted well and other tests only confirm it. In the code above, you can see however that the C code only flattens the two-dim array and passes the flattened array to XGBoost.

wendycwong commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java:
```diff
@@ -540,95 +533,90 @@ private static int initalizeFromChunkIds(Frame f, int[] chunks, Vec.Reader[] vec
+                    if (vecs[j].isNA(i)) {
+                        colIndex[currentRow][currentCol++] = di.getCategoricalId(j, Float.NaN);
+                    } else {
                         colIndex[currentRow][currentCol++] = di.getCategoricalId(j, vecs[j].at8(i));
```
This one is super nice here

Pscheidl commented on h2o-py/tests/testdir_algos/xgboost/pyunit_xgboost_zero_nan_handling_native_comparison.py:
```diff
@@ -43,15 +59,15 @@ def xgboost_categorical_zero_nan_handling_sparse_vs_native():
-    assert sparse_based_prediction['predict'][1,0] == native_prediction[1].item()
+    print(native_prediction)
+    print(sparse_based_prediction)
+    assert sparse_based_prediction['predict'][0,0] == sparse_based_prediction['predict'][1,0]
```
I test here that the two first predictions are the same (predicting NaN and Zero). They are indeed the same, but Native XGBoost will give a mean of the two values - 25, however H2O will give better result - 30. As ETA is set exactly to 1, it should immediately learn the value 30. In case of native XGBoost, that does not happen. Maybe this is a parameter issue. Spent quite some time trying to figure out what's wrong with parameters, found no difference.

However, it treats the both values the same way, which was the purpose of this test and the asserts test that.

wendycwong commented:
Pavel:

Your tests may fail from time to time unless you run them in single node mode.  To do this, you need to add NOPASS somewhere in your test name.  Also, add your tests to h2o-3/tests/pyunitSingleNodeList file.

Pscheidl commented:
- Rebased on top of master
- Addressed Wendy's point about not running the tests as single-node.
- Added mixed tests for sparse (both num & cat columns) with traditional variety of response variable type (two levels of categorical, multiple levels of categorical & numerical).
- Removed zero-padding in case of an empty row. I believe the reason this bad approach did not break anything is simple - the number of "non-zero" (there was an actual zero in the matrix) elements was raised by one, yet the element was zero and thus non-significant.
